hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/TargetRepository.java

```diff
@@ -115,164 +96,105 @@ void setAssignedAndInstalledDistributionSetAndUpdateStatus(@Param("status") Targ
+    @Override
+    @NonNull
+    default Page<JpaTarget> findAll(@NonNull Pageable pageable) {
+        return this.findAll(Specification.where(null), pageable);
+    }
 
     /**
-     * Counts number of targets with given distribution set Id.
+     * Finds {@link Target}s in the repository matching a list of ids.
      *
-     * @param distId
-     *            to search for
+     * Calls version based on spec to allow injecting further specs
      *
-     * @return number of found {@link Target}s.
+     * @param ids ids to filter for
+     * @return {@link List} of {@link Target}s
      */
-    Long countByAssignedDistributionSetId(Long distId);
+    @Override
+    @NonNull
+    default List<JpaTarget> findAllById(Iterable<Long> ids) {
+        final List<Long> collectedIds = StreamSupport.stream(ids.spliterator(), true).collect(Collectors.toList());
+        return this.findAll(Specification.where(TargetSpecifications.hasIdIn(collectedIds)));
+    }
 
     /**
-     * Counts number of targets with given distribution set Id.
+     * Finds {@link Target} in the repository matching an id.
      *
-     * @param distId
-     *            to search for
-     * @return number of found {@link Target}s.
-     */
-    Long countByInstalledDistributionSetId(Long distId);
-
-    /**
-     * Checks if there is already a {@link Target} that has the given distribution set Id assigned or installed.
+     * Calls version based on spec to allow injecting further specs
      *
-     * @param distId
-     *            to check
-     * @return <code>true</code> if a {@link Target} exists.
+     * @param id id to filter for
+     * @return {@link Optional} of {@link Target}
      */
-    @Query("SELECT CASE WHEN COUNT(t)>0 THEN 'true' ELSE 'false' END FROM JpaTarget t WHERE t.installedDistributionSet.id=:distId OR t.assignedDistributionSet.id=:distId")
-    boolean existsByInstalledOrAssignedDistributionSet(@Param("distId") Long distId);
+    @Override
+    @NonNull
+    default Optional<JpaTarget> findById(@NonNull Long id) {
+        return this.findOne(Specification.where(TargetSpecifications.hasId(id)));
+    }
 
     /**
-     * Finds all {@link Target}s in the repository.
+     * Checks whether {@link Target} in the repository matching an id exists or not.
      *
-     * @return {@link List} of {@link Target}s
+     * Calls version based on spec to allow injecting further specs
      *
-     * @see org.springframework.data.repository.CrudRepository#findAll()
+     * @param id id to check for
+     * @return true if target with id exists
      */
     @Override
-    List<JpaTarget> findAll();
-
-    @Override
-    // Workaround for https://bugs.eclipse.org/bugs/show_bug.cgi?id=349477
-    @Query("SELECT t FROM JpaTarget t WHERE t.id IN ?1")
-    List<JpaTarget> findAllById(Iterable<Long> ids);
+    default boolean existsById(@NonNull Long id) {
+        return this.findOne(Specification.where(TargetSpecifications.hasId(id))).isPresent();
```

Comment:
what about:
```suggestion
        return this.count(Specification.where(TargetSpecifications.hasId(id))) > 0;
```

Reply:
Is also covered by [last commit](https://github.com/eclipse/hawkbit/pull/1148/commits/565d5e8b4b479eb7f7efb8ddbb2103f094c995c6) to streamline existence checks with specs based on count.